README.md
Colir
=====

[![Build Status][ci-badge]][ci-link]

* Repository: [https://github.com/kyrylo/colir][cr]

Description
-----------

This tiny library provides support for RGB colours and some simple manipulations
with them.

Installation
------------

    gem install colir

Synopsis
--------

The name Colir means “colour” in Ukrainian (roughly). OK, enough of rant, let's
talk about the API instead.

###### Basics

You can create colours with human readable names. For example, the following
code would create opaque red colour (0xff000000).

```ruby
require 'colir'

# Create a new Colir.
red = Colir.red #=> #<Colir:...>
# Get a valid RGB colour, represented as an integer number.
red.hex #=> 16711680
# Why do you never trust me?
red.hex.to_s(16) #=> "ff0000"

# You can also get the same colour, but with alpha channel.
red.hexa #=> 4278190080
# Okay, okay!
red.hexa.to_s(16) #=> "ff000000"
```

The list of all human readable colour names can be found at [w3schools][w3].
Actually, you can create any valid RGB colours, not just the predefined ones.

```ruby
# This is valid.
yup = Colir.new(0xff0000) #=> #<Colir:...>
# This isn't. There is no such RGB colour.
nope = Colir.new(0xff00001) #=> RangeError: out of allowed RGB values
```

###### Transparency

As you may have noticed, the library supports transparent colours.

```ruby
# Shortcut methods have only 1 parameter, which is transparency. It must lie
# within the range of 0..1
red = Colir.red(0.3)
red.hexa #=> 4278190110
red.hexa.to_s(16) #=> "ff00001e"
red.transparency #=> 0.3

# With arbitrary colours you can pass the second parameter.
green = Colir.new(0x00ff00, 0.3)
green.hexa #=> 16711710
green.hexa.to_s(16) #=> "0x00ff001e"
green.transparency #=> 0.3

# The default transparency is 0.0
green = Colir.new(0x00ff00)
green.transparency #=> 0.0
```

It's possible to adjust a Colir's transparency.

```ruby
# `#opaque!` and `#transparent!` set a Colir's transparency to `0.0`
# and `1.0` respectively.
blue = Colir.new(0x0000ff, 0.45)
blue.transparency #=> 0.3
blue.transparent! #=> (Colir: 0x00ff00e1)
blue.transparency #=> 0.0
blue.opaque! #=> (Colir: 0x00ff00ff)
blue.transparency #=> 1.0

# You can set your own transparency value.
blue.transparency = 0.9 #=> 0.9
blue.transparency #=> 0.9

# But be careful, as the valid value is anything in between 0 and 1.
blue.transparency = 1.01 #=> RangeError: ...
```

###### Shades

Last but not least, you can use colour shades.

```ruby
yellow = Colir.yellow
yellow.shade #=> 0

# Let's make it a bit darker.
yellow.darken #=> #<Colir:...>
yellow.shade #=> -1
yellow.hex #=> 13421568

# Hum, that's not dark enough. Let's do it one more time.
yellow.darken
yellow.shade #=> -2
yellow.hex #=> 10066176

# Actually, the previous shade was better!
yellow.lighten
yellow.shade #=> -1
yellow.hex #=> 13421568
```

The previous examples modify the Colir object (`#darken` and `#lighten`).
Hovewer, there are two other methods that return a new object: `#darker` and
`#lighter`.

```ruby
# HEXes
orange = Colir.orange
orange.hex #=> 16753920
orange.darker.hex #=> 13403392
orange.hex #=> 16753920

# Shades
orange = Colir.orange
orange.shade #=> 0
orange.lighter.shade #=> 1
orange.shade #=> 0
```

There is a handy way to reset the shade level. Can you guess it? Of course, it's
the `#reset_shade` method!

```ruby
indigo = Colir.indigo
indigo.lighten
indigo.lighten
indigo.lighten
indigo.lighten
indigo.lighten
indigo.shades #=> 5
indigo.reset_shade #=> #<Colir:...>
indigo.shade #=> 0
```

Limitations
-----------

### OS support

Colir is a cross-platform library.

### Rubies

* Ruby 1.9.3 and higher
* Rubinius 2.0.0rc1 and higher (mode 1.9 and higher)

Initially I wanted to support JRuby 1.7.3 (mode 1.9), but it has a [critical bug
in `BigDecimal` library](http://jira.codehaus.org/browse/JRUBY-7101), which
Colir actively utilises.

Credits
-------

* The first contributor badge goes to [kachick][kachick]

Licence
-------

The project uses Zlib licence. See LICENCE file for more information.

[cr]: https://github.com/kyrylo/colir
[ci-badge]: https://travis-ci.org/kyrylo/colir.png?branch=master "Build status"
[ci-link]: https://travis-ci.org/kyrylo/colir/ "Build history"
[w3]: http://www.w3schools.com/cssref/css_colornames.asp
[kachick]: https://github.com/kachick


CHANGELOG.md
Colir changelog
===============

### v0.0.1 (April 05, 2013)

* Initial release


spec/helper.rb
require 'bacon'
require 'pry'

require_relative '../lib/colir'

puts "Ruby: #{ RUBY_VERSION }; Colir version: #{ Colir::VERSION }"

def kind_of(klass)
  lambda { |obj| obj.kind_of?(klass) }
end


spec/colir_spec.rb
require_relative 'helper'

describe Colir do
  describe "some shortcuts" do
    it "has a shortcut for red colour" do
      Colir.red.hexa.should == 0xff000000
    end

    it "has a shortcut for blanched almond colour" do
      Colir.blanched_almond.hexa.should == 0xffebcd00
    end

    it "has a shortcut royal blue colour" do
      Colir.royal_blue.hexa.should == 0x4169e100
    end

    it "can set the colour's transparency" do
      Colir.snow(0.5).transparency.should == 0.5
    end
  end

  describe "#intialize" do
    it "accepts transparency parameter" do
      Colir.new(0x012345, 0.77).transparency.should == 0.77
    end

    it "doesn't allow too low HEX number" do
      should.raise(RangeError) {
        Colir.new(-0x000001)
      }.message.should =~ /out of allowed RGB values/
    end

    it "doesn't allow too high HEX number" do
      should.raise(RangeError) {
        Colir.new(0x1000000)
      }.message.should =~ /out of allowed RGB values/
    end
  end

  describe "#hex" do
    it "returns the hex code of a colour (without transparency)" do
      Colir.new(0x123456, 0.1).hex.should == 0x123456
    end
  end

  describe "#hexa" do
    it "returns the hex code of a colour (with transparency)" do
      Colir.new(0x123456, 0.1).hexa.should == 0x1234560a
    end
  end

  describe "#transparency" do
    before do
      @colir = Colir.new(0x123456)
    end

    it "defaults to no transparency" do
      @colir.transparency.should == 0.0
    end

    it "can be set to a specific value" do
      Colir.new(0x123456, 0.98).transparency.should == 0.98
    end

    it "must lie within the range of 0 to 1 (min)" do
      should.raise(RangeError) {
        Colir.new(0x123456, -0.01)
      }.message.should =~ /out of allowed transparency values \(0-1\)/
    end

    it "must lie within the range of 0 to 1 (max)" do
      should.raise(RangeError) {
        Colir.new(0x123456, 1.01)
      }.message.should =~ /out of allowed transparency values \(0-1\)/
    end

    it "returns a Float value" do
      @colir.transparency.should.be kind_of(Float)
    end
  end

  describe "#transparency=" do
    before do
      @colir = Colir.new(0x123456)
    end

    it "can be set" do
      @colir.transparency = 0.89
      @colir.transparency.should == 0.89
    end

    it "raises error on a bad value" do
      should.raise(RangeError) {
        @colir.transparency = 1.01
      }.message.should =~ /out of allowed transparency values/
    end
  end

  describe "#transparent!" do
    it "changes the object's transparency to full transparency" do
      colir = Colir.new(0x123456, 0.3).transparent!
      colir.transparency.should == 1.0
    end
  end

  describe "#opaque!" do
    it "changes the object's transparency to no transparency" do
      colir = Colir.new(0x123456, 0.3).opaque!
      colir.transparency.should == 0.0
    end
  end

  describe "#shade" do
    before do
      @colir = Colir.new(0x123456)
    end

    it "returns 0 colour shade for a colour without shades" do
      @colir.shade.should == 0
    end

    it "returns -1 for a darker colour" do
      @colir.darken.shade.should == -1
    end

    it "returns 1 for a lighter colour" do
      @colir.lighten.shade.should == 1
    end
  end

  describe "#reset_shade" do
    before do
      @colir = Colir.new(0x123456)
    end

    it "resets the colour to its initial state" do
      before = @colir.shade
      3.times { @colir.darken }
      @colir.shade.should.not == before
      @colir.reset_shade
      @colir.shade.should == before
    end

    it "restores the hex number" do
      before = @colir.hex
      3.times { @colir.darken }
      @colir.hex.should.not == before
      @colir.reset_shade
      @colir.hex.should == before
    end
  end

  describe "#darker" do
    before do
      @colir = Colir.new(0x123456)
    end

    it "returns a new object" do
      before_id = @colir.object_id
      after_id = @colir.darker.object_id
      after_id.should.not == before_id
    end

    it "doesn't modify the base colour" do
      before = @colir.hex
      @colir.darker
      @colir.hex.should == before
    end

    it "returns a new colour, but a little bit darker" do
      @colir.darker.shade.should == -1
    end
  end

  describe "#lighter" do
    before do
      @colir = Colir.new(0x123456)
    end

    it "returns a new object" do
      before_id = @colir.object_id
      after_id = @colir.lighter.object_id
      after_id.should.not == before_id
    end

    it "doesn't modify the base colour" do
      before = @colir.hex
      @colir.lighter
      @colir.hex.should == before
    end

    it "returns a new colour, but a little bit darker" do
      @colir.lighter.shade.should == 1
    end
  end

  describe "#darken" do
    before do
      @colir = Colir.new(0x123456)
    end

    it "modifies the base colour" do
      old_colour = @colir.hex
      @colir.darken
      @colir.hex.should.not == old_colour
    end

    it "makes the current colour darker" do
      @colir.darken
      @colir.hex.should == 0x0e2a45
    end

    it "makes the current colour even more dark" do
      3.times { @colir.darken }
      @colir.hex.should == 0x071522
    end

    describe "lower limit" do
      before do
        5.times { @colir.darken }
      end

      it "stops darkening the current colour when meets the limit of shades" do
        old_hex = @colir.hex
        @colir.darken.hex.should == old_hex
      end

      it "finally reaches black colour" do
        @colir.hex.should == 0x000000
      end
    end
  end

  describe "#lighten" do
    before do
      @colir = Colir.new(0x123456)
    end

    it "modifies the base colour" do
      old_colour = @colir.hex
      @colir.lighten
      @colir.hex.should.not == old_colour
    end

    it "makes the current colour lighter" do
      @colir.lighten
      @colir.hex.should == 0x205d99
    end

    it "makes the current colour even more light" do
      3.times { @colir.lighten }
      @colir.hex.should == 0x79aee3
    end

    describe "upper limit" do
      before do
        5.times { @colir.lighten }
      end

      it "stops lightening the current colour when meets the limit of shades" do
        old_hex = @colir.hex
        @colir.lighten.hex.should == old_hex
      end

      it "finally reaches white colour" do
        @colir.hex.should == 0xffffff
      end
    end
  end

  describe "shading" do
    before do
      @colir = Colir.new(0x123456)
    end

    it "darkens properly" do
      @colir.darken.hex.should == 0x0e2a45
      @colir.darken.hex.should == 0x0b1f34
      @colir.darken.hex.should == 0x071522
      @colir.darken.hex.should == 0x040a11
      @colir.darken.hex.should == 0
    end

    it "lightens properly" do
      @colir.lighten.hex.should == 0x205d99
      @colir.lighten.hex.should == 0x3685d5
      @colir.lighten.hex.should == 0x79aee3
      @colir.lighten.hex.should == 0xbcd6f1
      @colir.lighten.hex.should == 0xffffff
    end

    it "darkens and lightens properly" do
      before = @colir.hex
      2.times { @colir.lighten }
      5.times { @colir.darken }
      10.times { @colir.lighten }
      5.times { @colir.darken }
      after = @colir.hex
      before.should == after
    end

    describe "lower limit - 0x000000" do
      describe "darkening" do
        before do
          @colir = Colir.new(0x000000)
        end

        it "doesn't change the hex value" do
          before = @colir.hex
          after = @colir.darken.hex
          after.should == before
        end

        it "does change the shade level" do
          @colir.darken.shade.should == -1
        end

        it "doesn't overflow the shade level" do
          6.times { @colir.darken }
          @colir.shade.should == -5
        end
      end

      describe "lightening" do
        before do
          @colir = Colir.new(0x000000)
        end

        it "does change the hex value" do
          before = @colir.hex
          after = @colir.lighten.hex
          after.should.not == before
        end

        it "does change the shade level" do
          @colir.lighten.shade.should == 1
        end

        it "doesn't overflow the shade level" do
          6.times { @colir.lighten }
          @colir.shade.should == 5
        end
      end
    end

    describe "very close to lower limit" do
      describe "darkening" do
        before do
          @colir = Colir.new(0x000001)
        end

        it "does change the hex value when it's time" do
          before = @colir.hex
          3.times { @colir.darken }
          after = @colir.hex
          after.should.not == before
        end

        it "stops darkening when reaches the lower limit" do
          3.times { @colir.darken }
          before = @colir.hex
          after = @colir.darken.hex
          after.should == before
        end

        it "keeps changing the shade level" do
          2.times { @colir.darken }
          @colir.shade.should == -2

          2.times { @colir.darken }
          @colir.shade.should == -4
        end
      end
    end

    describe "very close to upper limit" do
      describe "lightening" do
        before do
          @colir = Colir.new(0xfffffe)
        end

        it "does change the hex value when it's time" do
          before = @colir.hex
          3.times { @colir.lighten }
          after = @colir.hex
          after.should.not == before
        end

        it "stops lightening when reaches the upper limit" do
          3.times { @colir.lighten }
          before = @colir.hex
          after = @colir.lighten.hex
          after.should == before
        end

        it "keeps changing the shade level" do
          2.times { @colir.lighten }
          @colir.shade.should == 2

          2.times { @colir.lighten }
          @colir.shade.should == 4
        end
      end
    end

  end
end


spec/hslrgb_spec.rb
require_relative 'helper'

describe Colir::HSLRGB do

  describe "::rgb_to_hsl" do
    it "converts black RGB colour to its HSL equivalent" do
      Colir::HSLRGB.rgb_to_hsl(0x00, 0x00, 0x00).should == [0, 0, 0]
    end

    it "converts white RGB colour to its HSL equivalent" do
      Colir::HSLRGB.rgb_to_hsl(0xff, 0xff, 0xff).should == [0, 0, 1]
    end

    it "converts reddish RGB colour to its HSL equivalent" do
      Colir::HSLRGB.rgb_to_hsl(0x99, 0x00, 0x00).should == [0, 1.0, 0.3]
    end

    it "converts orangish RGB colour to its HSL equivalent" do
      hsl = Colir::HSLRGB.rgb_to_hsl(0xff, 0xa3, 0x19)
      hsl[0].should == 36
      hsl[1].should == 1.0
      hsl[2].round(2).should == 0.55
    end

    it "converts yellowish RGB colour to its HSL equivalent" do
      hsl = Colir::HSLRGB.rgb_to_hsl(0xff, 0xff, 0x19)
      hsl[0].should == 60
      hsl[1].should == 1.0
      hsl[2].round(2).should == 0.55
    end

    it "converts greenish RGB colour to its HSL equivalent" do
      Colir::HSLRGB.rgb_to_hsl(0x33, 0xcc, 0x33).should == [120, 0.6, 0.5]
    end

    it "converts bluish RGB colour to its HSL equivalent" do
      hsl = Colir::HSLRGB.rgb_to_hsl(0x19, 0xff, 0xff)
      hsl[0].should == 180
      hsl[1].should == 1.0
      hsl[2].round(2) == 0.55
    end

    it "converts indigish RGB colour to its HSL equivalent" do
      Colir::HSLRGB.rgb_to_hsl(0x00, 0x66, 0xff).should == [216, 1.0, 0.5]
    end

    it "converts violetish RGB colour to its HSL equivalent" do
      Colir::HSLRGB.rgb_to_hsl(0x99, 0x33, 0xff).should == [270, 1.0, 0.6]
    end

    describe "return value" do
      before do
        @hsl = Colir::HSLRGB.rgb_to_hsl(0xaf, 0xfe, 0x39)
      end

      it "should be Array" do
        @hsl.should.be kind_of(Array)
      end

      it "returns an integer hue" do
        @hsl[0].should.be kind_of(Integer)
      end

      it "returns a BigDecimal saturation" do
        @hsl[1].should.be kind_of(BigDecimal)
      end

      it "returns a BigDecimal lightness" do
        @hsl[2].should.be kind_of(BigDecimal)
      end
    end

    describe "parameters" do
      describe "R" do
        it "raises RangeError if it's too low" do
          should.raise(RangeError) {
            Colir::HSLRGB.rgb_to_hsl(-1, 0, 0)
          }.message =~ /out of allowed RGB values \(0-255\)/
        end

        it "raises RangeError if it's too high" do
          should.raise(RangeError) {
            Colir::HSLRGB.rgb_to_hsl(256, 255, 255)
          }.message =~ /out of allowed RGB values \(0-255\)/
        end
      end

      describe "G" do
        it "raises RangeError if it's too low" do
          should.raise(RangeError) {
            Colir::HSLRGB.rgb_to_hsl(0, -1, 0)
          }.message =~ /out of allowed RGB values \(0-255\)/
        end

        it "raises RangeError if it's too high" do
          should.raise(RangeError) {
            Colir::HSLRGB.rgb_to_hsl(255, 256, 255)
          }.message =~ /out of allowed RGB values \(0-255\)/
        end

      end

      describe "B" do
        it "raises RangeError if it's too low" do
          should.raise(RangeError) {
            Colir::HSLRGB.rgb_to_hsl(0, 0, -1)
          }.message =~ /out of allowed RGB values \(0-255\)/
        end

        it "raises RangeError if it's too high" do
          should.raise(RangeError) {
            Colir::HSLRGB.rgb_to_hsl(255, 255, 256)
          }.message =~ /out of allowed RGB values \(0-255\)/
        end
      end
    end
  end

  describe "::hsl_to_rgb" do
    it "converts black HSL colour to its RGB equivalent" do
      Colir::HSLRGB.hsl_to_rgb(0, 0, 0).should == [0, 0, 0]
    end

    it "converts white HSL colour to its RGB equivalent" do
      Colir::HSLRGB.hsl_to_rgb(0, 0, 1.0).should == [0xff, 0xff, 0xff]
    end

    it "converts greyish HSL colour to its RGB equivalent" do
      Colir::HSLRGB.hsl_to_rgb(360, 0, 0.8).should == [0xcc, 0xcc, 0xcc]
    end

    it "converts reddish HSL colour to its RGB equivalent" do
      Colir::HSLRGB.hsl_to_rgb(0, 1.0, 0.3).should == [0x99, 0x00, 0x00]
    end

    it "converts orangish HSL colour to its RGB equivalent" do
      Colir::HSLRGB.hsl_to_rgb(36, 1.0, 0.55).should == [0xff, 0xa3, 0x1a]
    end

    it "converts yellowish HSL colour to its RGB equivalent" do
      Colir::HSLRGB.hsl_to_rgb(60, 1.0, 0.55).should == [0xff, 0xff, 0x1a]
    end

    it "converts greenish HSL colour to its RGB equivalent" do
      Colir::HSLRGB.hsl_to_rgb(120, 0.6, 0.5).should == [0x33, 0xcc, 0x33]
    end

    it "converts bluish HSL colour to its RGB equivalent" do
      Colir::HSLRGB.hsl_to_rgb(180, 1.0, 0.55).should == [0x1a, 0xff, 0xff]
    end

    it "converts indigish HSL colour to its RGB equivalent" do
      Colir::HSLRGB.hsl_to_rgb(216, 1.0, 0.5).should == [0x00, 0x66, 0xff]
    end

    it "converts violetish HSL colour to its RGB equivalent" do
      Colir::HSLRGB.hsl_to_rgb(270, 1.0, 0.6).should == [0x99, 0x33, 0xff]
    end

    describe "parameters" do
      describe "H" do
        it "raises RangeError if it's too low" do
          should.raise(RangeError) {
            Colir::HSLRGB.hsl_to_rgb(-1, 0, 0)
          }.message =~ /out of allowed HSL values/
        end

        it "raises RangeError if it's too high" do
          should.raise(RangeError) {
            Colir::HSLRGB.hsl_to_rgb(361, 1, 1)
          }.message =~ /out of allowed HSL values/
        end
      end

      describe "S" do
        it "raises RangeError if it's too low" do
          should.raise(RangeError) {
            Colir::HSLRGB.hsl_to_rgb(0, -0.1, 0)
          }.message =~ /out of allowed HSL values/
        end

        it "raises RangeError if it's too high" do
          should.raise(RangeError) {
            Colir::HSLRGB.hsl_to_rgb(360, 1.1, 1)
          }.message =~ /out of allowed HSL values/
        end

      end

      describe "L" do
        it "raises RangeError if it's too low" do
          should.raise(RangeError) {
            Colir::HSLRGB.hsl_to_rgb(0, 1, -0.1)
          }.message =~ /out of allowed HSL values/
        end

        it "raises RangeError if it's too high" do
          should.raise(RangeError) {
            Colir::HSLRGB.hsl_to_rgb(360, 1, 1.1)
          }.message =~ /out of allowed HSL values/
        end
      end
    end
  end
end



lib/colir/hslrgb.rb
require 'bigdecimal'

class Colir
  # The module provides methods for RGB to HSL and HSL to RGB conversions. I
  # just ported it from a C implementation.
  # @see http://goo.gl/WLwi6
  # @since 0.0.1
  # @private
  module HSLRGB

    # Provides helper methods for the RGB colours.
    module RGB

      # The possible values of an RGB colour.
      RGB_RANGE = 0x000000..0xffffff

      # The possible values of an RGB byte.
      RGB_BYTE_RANGE = 0x00..0xff

      # Performs a validation check for the +rgb+.
      #
      # @example Truthy
      #   RGB.valid_rgb?([255, 13, 0]) #=> true
      #
      # @example Falsey
      #   RGB.valid_rgb?([256, 13, -1]) #=> false
      #
      # @param [Array<Integer>] rgb The RGB colour to be checked
      # @return [Boolean] true if the given +rgb+ colour lies within the
      #   `RGB_BYTE_RANGE`
      def self.valid_rgb?(rgb)
        rgb.all? { |b| valid_byte?(b) } && (rgb.length == 3)
      end

      # Converts +hex+ number to the RGB array.
      #
      # @example
      #   RGB.int_bytes(0x123456) #=> [18, 52, 86]
      #
      # @param [String] hex The hex number expressed without the preceding `0x`
      # @return [Array<Integer>] the RGB array
      def self.int_bytes(hex)
        hex.to_s(16).rjust(6, '0').scan(/../).map { |b| b.to_i(16) }
      end

      def self.valid_byte?(byte)
        RGB_BYTE_RANGE.cover?(byte) && byte.integer?
      end
      private_class_method :valid_byte?
    end

    # Provides helper methods for HSL colours.
    module HSL

      # The possible values for the hue.
      H_RANGE = 0..360

      # The possible values for the saturation.
      S_RANGE = 0..1

      # The possible values for the lightness.
      L_RANGE = S_RANGE

      # Performs a validation check for the +hsl+.
      #
      # @example Truthy
      #   RGB.valid_hsl?([180, 1, 0.55]) #=> true
      #
      # @example Falsey
      #   RGB.valid_hsl?([180, 1.1, 0.55]) #=> false
      #
      # @param [Array<Number>] hsl The HSL colour to be checked
      # @return [Boolean] true if the given +hsl+ colour lies within the
      #   `H_RANGE`, `S_RANGE`, `L_RANGE`
      def self.valid_hsl?(hsl)
        valid_hue?(hsl[0]) && S_RANGE.cover?(hsl[1]) &&
          L_RANGE.cover?(hsl[2]) && (hsl.length == 3)
      end

      def self.valid_hue?(hue)
        H_RANGE.cover?(hue) && hue.integer?
      end
      private_class_method :valid_hue?
    end

    # The cached array of arrays, which contains order codes for the model  from
    # `::hsl_to_rgb`. The model is an array, where the needed for conversion
    # values are calculated only once. The order codes allows to refer to the
    # calculated model array values via indices. Hopefully, it saves some CPU
    # cycles.
    RGB_ORDER_CODES = [
      [0, 1, 2], # [0, 60)
      [1, 0, 2], # [60, 120)
      [2, 0, 1], # [120, 180)
      [2, 1, 0], # [180, 240)
      [1, 2, 0], # [240, 300)
      [0, 2, 1], # [300, 360)
    ]

    # The code returned for 0 or 360 degrees.
    RGB_DEFAULT_CODE = [2, 2, 2]

    # Converts an RGB colour to its HSL counterpart.
    #
    # @example
    #   rgb = 'c0c0c0'.scan(/../).map { |b| b.to_i(16) } #=> [192, 192, 192]
    #   HSLRGB.rgb_to_hsl(192, 192, 192) #=> [0, 0.0, 0.75]
    #
    # @param [Integer] red Possible values: 0..255
    # @param [Integer] green Possible values: 0..255
    # @param [Integer] blue Possible values: 0..255
    # @return [Array<Integer,BigDecimal>] the converted HSL representation of
    #   an RGB colour
    # @raise [RangeError] if one of the parameters doesn't lie within the
    #   accepted range
    # @see http://en.wikipedia.org/wiki/HSL_color_space HSL color space
    # @see ::hsl_to_rgb
    def self.rgb_to_hsl(red, green, blue)
      validate_rgb!([red, green, blue])

      red, green, blue = [red, green, blue].map { |b| b / BigDecimal('255.0') }
      min, max = [red, green, blue].minmax
      chroma = max - min

      lightness = (min + max) * BigDecimal('0.5')

      hue = 0
      saturation = BigDecimal('0.0')

      if chroma.nonzero?
        hue = case max
              when red   then (green - blue) / chroma % 6
              when green then (blue - red) / chroma + 2
              else            (red - green) / chroma + 4
              end
        hue = (hue * 60).round
        saturation = chroma / (1 - (2 * lightness - 1).abs)
      end

      [hue, saturation, lightness]
    end

    def self.validate_rgb!(rgb)
      unless RGB.valid_rgb?(rgb)
        raise RangeError, 'out of allowed RGB byte values (0-255)'
      end
    end
    private_class_method :validate_rgb!

    def self.validate_hsl!(hsl)
      unless HSL.valid_hsl?(hsl)
        raise RangeError, 'out of allowed HSL values (h:0-360 s:0-1 l:0-1)'
      end
    end
    private_class_method :validate_hsl!

    # Converts an HSL colour to its RGB counterpart. The algorithm is correct,
    # but sometimes you'll notice little laxity (because of rounding problems).
    #
    # @example
    #   HSLRGB.hsl_to_rgb(180, 1, 0.55) #=> [26, 255, 255]
    #
    # @param [Integer] hue Possible values: 0..360
    # @param [BigDecimal] saturation Possible values: 0..1
    # @param [BigDecimal] lightness Possible values: 0..1
    # @return [Array<Integer>] the converted RGB representation of a HSL colour
    # @raise [RangeError] if one of the parameters doesn't lie within the
    #   accepted range
    # @see ::rgb_to_hsl
    def self.hsl_to_rgb(hue, saturation, lightness)
      validate_hsl!([hue, saturation, lightness])

      chroma = (1 - (2 * lightness - 1).abs) * saturation
      a = 1 * (lightness - 0.5 * chroma)
      b = chroma * (1 - ((hue / BigDecimal('60.0')).modulo(2) - 1).abs)

      degrees = Array.new(7) { |idx| idx * 60 }.each
      model = [chroma+a, a+b, a]

      RGB_ORDER_CODES.find(->{RGB_DEFAULT_CODE}) {
        (degrees.next...degrees.peek) === hue
      }.map { |id|
        (model[id] * 255).round
      }
    end

  end
end
